File: Part 2 - Regression/Section 8 - Decision Tree Regression/Decision_Tree_reg.py
# Regression Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


#As this is regression problem we will use decision tree regressor
#We can also opt for the classifier depending on the problem 

from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(X,y)


# Predicting a new result
y_pred = regressor.predict([6.5])

# Visualising the Regression results
#As we can check that it is not showing best of results 
#As the regressor divides it into considering each level as one class. 

plt.scatter(X, y, color = 'red')
plt.plot(X, regressor.predict(X), color = 'blue')
plt.title('Truth or Bluff (Regression Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

#By increasing the number of intervals we can check it is flat on each level.
# Visualising the Regression results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Truth or Bluff (Regression Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: Part 2 - Regression/Section 8 - Decision Tree Regression/Decision_Tree_reg.R
# Regression Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]
#Installing the package for the decision tree 
install.packages('rpart')
#including the library 
library('rpart')


#building the decision tree regressor

regressor = rpart(formula = dataset$Salary ~ .,dataset)

# Predicting a new result
#The results are not good as the decision tree is making only one split :( 

y_pred = predict(regressor, data.frame(Level = 6.5))


#we will use rpart to add min number of splits there should be 

regressor = rpart(formula = dataset$Salary ~ .,dataset,
                  control = rpart.control(minsplit = 5))

# Predicting a new result
#The results are not good as the decision tree is making only one split :( 

y_pred = predict(regressor, data.frame(Level = 6.5))



# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Descison Regression Model)') +
  xlab('Level') +
  ylab('Salary')